codebase2/codebase-sqlite/U/Codebase/Sqlite/Reference.hs

```diff
@@ -27,15 +29,12 @@ type IdH = Id' HashId
 
 -- * Orphan instances
 
-instance ToRow (Reference' TextId HashId) where
+instance (ToField t, ToField h) => ToRow (Reference' t h) where
   toRow = \case
     ReferenceBuiltin t -> toRow (Only t) ++ [SQLNull, SQLNull]
     ReferenceDerived (Id h i) -> SQLNull : toRow (Only h) ++ toRow (Only i)
 
-instance FromRow (Reference' TextId HashId) where
-  fromRow = referenceFromRow'
-
-instance FromRow (Reference' TextId ObjectId) where
+instance (FromField t, FromField h, Show t, Show h) => FromRow (Reference' t h) where
```

Comment:
I'm thinking it might be better to have explicit instances for the specific kinds of References we find in the database, and not every combination like `Reference' UTCTime Float`

Reply:
You can use the `referenceFromRow'` trick like `referenceToRow'` that I'd missed, to cut code duplication.